# «Облегчённые» матрицы потерь: 13 альтернатив

Используется пакет работы с ранжированиями `Rankings.jl`. Функция `FindMinima()` последовательного перебора всех перестановок из файла `FindScoreMatrixMinima.jl` будет здесь бесполезна -- из-за слишком большого их количества.

In [1]:
using Rankings

## Муравьёв (2013)

S.V. Muravyov. **Ordinal measurement, preference aggregation and interlaboratory comparisons.** *Measurement,* **46** (2013) 2927--2935. ([PDF](https://portal.tpu.ru/f_ic/files/international/publications/27.pdf))

В статье альтернативы представлены лишь своими порядковыми номерами (см. страницу 2932, таблица 5); здесь они
обозначаются символом `a` с соответствующими нижними индексами, поскольку
в используемых пакетом `Rankings.jl` соглашениях о представлении ранжирований
только цифрами могут быть записаны не более девяти ранжирований, причём не связанных.

In [2]:
Муравьёв2013_2932_T5 = ranking"a₁₀~a₁₁ a₀₁~a₀₂~a₀₃~a₀₄~a₀₅~a₀₆~a₀₇~a₀₈~a₀₉~a₁₂~a₁₃"⦸
                       ranking"a₁₀ a₀₁~a₀₂~a₀₃~a₀₄~a₀₅~a₀₆~a₀₇~a₀₈~a₀₉~a₁₁~a₁₂~a₁₃"⦸
                       ranking"a₁₀~a₁₁ a₀₁~a₀₂~a₀₃~a₀₄~a₀₅~a₀₆~a₀₇~a₀₈~a₀₉~a₁₂~a₁₃"⦸
                       ranking"a₀₉~a₁₀~a₁₁ a₀₁~a₀₂~a₀₃~a₀₄~a₀₅~a₀₆~a₀₇~a₀₈~a₁₂~a₁₃"⦸
                       ranking"a₀₈~a₀₉~a₁₀~a₁₁ a₁₂~a₀₁~a₀₂~a₀₃~a₀₄~a₀₅~a₀₆~a₀₇~a₁₃"⦸
                       ranking"a₀₈~a₀₉~a₁₀ a₀₁~a₀₂~a₀₃~a₀₄~a₀₅~a₀₆~a₀₇~a₁₁~a₁₂~a₁₃"⦸
                       ranking"a₁₁~a₁₂ a₀₁~a₀₂~a₀₃~a₀₄~a₀₅~a₀₆~a₀₇~a₀₈~a₀₉~a₁₀~a₁₃"⦸
                       ranking"a₀₉~a₁₀~a₁₁ a₀₁~a₀₂~a₀₃~a₀₄~a₀₅~a₀₆~a₀₇~a₀₈~a₁₂~a₁₃"⦸
                       ranking"a₀₂~a₀₃ a₀₁~a₀₄~a₀₅~a₀₆~a₀₇~a₀₈~a₀₉~a₁₀~a₁₁~a₁₂~a₁₃"⦸
                       ranking"a₀₈~a₀₉ a₀₁~a₀₂~a₀₃~a₀₄~a₀₅~a₀₆~a₀₇~a₁₀~a₁₁~a₁₂~a₁₃"

13×13 Array{Int64,2}:
  0   1   1   0   0   0   0   3   5  7   6   1   0
 -1   0   0  -1  -1  -1  -1   2   4  6   5   0  -1
 -1   0   0  -1  -1  -1  -1   2   4  6   5   0  -1
  0   1   1   0   0   0   0   3   5  7   6   1   0
  0   1   1   0   0   0   0   3   5  7   6   1   0
  0   1   1   0   0   0   0   3   5  7   6   1   0
  0   1   1   0   0   0   0   3   5  7   6   1   0
 -3  -2  -2  -3  -3  -3  -3   0   2  4   3  -2  -3
 -5  -4  -4  -5  -5  -5  -5  -2   0  2   1  -4  -5
 -7  -6  -6  -7  -7  -7  -7  -4  -2  0  -1  -6  -7
 -6  -5  -5  -6  -6  -6  -6  -3  -1  1   0  -5  -6
 -1   0   0  -1  -1  -1  -1   2   4  6   5   0  -1
  0   1   1   0   0   0   0   3   5  7   6   1   0

In [3]:
consensus(Муравьёв2013_2932_T5)

┌ Info: Sum of negative values is -208.
└ @ Rankings ~/.julia/packages/Rankings/GfcmE/src/consensus.jl:192


Δ⁺(score_mat) = 0


┌ Info: col_optimize() stopped at [10, 11, 9, 8, 2, 3, 12, 1, 4, 5, 6, 7, 13].
└ @ Rankings ~/.julia/packages/Rankings/GfcmE/src/consensus.jl:165
┌ Info: Upper-triangle sum is -208.
└ @ Rankings ~/.julia/packages/Rankings/GfcmE/src/consensus.jl:166
┌ Info: Collecting equivalent permutations...
└ @ Rankings ~/.julia/packages/Rankings/GfcmE/src/consensus.jl:55
┌ Info: row_optimize() stopped at [10, 11, 9, 8, 12, 3, 2, 13, 7, 6, 5, 4, 1].
└ @ Rankings ~/.julia/packages/Rankings/GfcmE/src/consensus.jl:183
┌ Info: Upper-triangle sum is -208.
└ @ Rankings ~/.julia/packages/Rankings/GfcmE/src/consensus.jl:184
┌ Info: Collecting equivalent permutations...
└ @ Rankings ~/.julia/packages/Rankings/GfcmE/src/consensus.jl:55
┌ Info: Upper-triangle sum reached: -208
└ @ Rankings ~/.julia/packages/Rankings/GfcmE/src/consensus.jl:212


4320-element Array{Array{Int64,1},1}:
 [10, 11, 9, 8, 2, 3, 12, 1, 4, 5, 6, 7, 13]
 [10, 11, 9, 8, 2, 3, 12, 1, 4, 5, 6, 13, 7]
 [10, 11, 9, 8, 2, 3, 12, 1, 4, 5, 7, 6, 13]
 [10, 11, 9, 8, 2, 3, 12, 1, 4, 5, 7, 13, 6]
 [10, 11, 9, 8, 2, 3, 12, 1, 4, 5, 13, 6, 7]
 [10, 11, 9, 8, 2, 3, 12, 1, 4, 5, 13, 7, 6]
 [10, 11, 9, 8, 2, 3, 12, 1, 4, 6, 5, 7, 13]
 [10, 11, 9, 8, 2, 3, 12, 1, 4, 6, 5, 13, 7]
 [10, 11, 9, 8, 2, 3, 12, 1, 4, 6, 7, 5, 13]
 [10, 11, 9, 8, 2, 3, 12, 1, 4, 6, 7, 13, 5]
 [10, 11, 9, 8, 2, 3, 12, 1, 4, 6, 13, 5, 7]
 [10, 11, 9, 8, 2, 3, 12, 1, 4, 6, 13, 7, 5]
 [10, 11, 9, 8, 2, 3, 12, 1, 4, 7, 5, 6, 13]
 ⋮
 [10, 11, 9, 8, 12, 3, 2, 13, 7, 5, 1, 4, 6]
 [10, 11, 9, 8, 12, 3, 2, 13, 7, 5, 1, 6, 4]
 [10, 11, 9, 8, 12, 3, 2, 13, 7, 5, 4, 1, 6]
 [10, 11, 9, 8, 12, 3, 2, 13, 7, 5, 4, 6, 1]
 [10, 11, 9, 8, 12, 3, 2, 13, 7, 5, 6, 1, 4]
 [10, 11, 9, 8, 12, 3, 2, 13, 7, 5, 6, 4, 1]
 [10, 11, 9, 8, 12, 3, 2, 13, 7, 6, 1, 4, 5]
 [10, 11, 9, 8, 12, 3, 2, 13, 7, 6, 1, 5, 4]
 [10, 11, 9, 8

Таким образом, нашлись 4320 перестановок (а не 1440, как указано в статье).
Структура всех 4320 оптимальных ранжирований одинакова: первыми всегда будут
альтернативы 10, 11, 9, 8 -- в указанном порядке. Затем (на следующих трёх позициях)
следуют все перестановки альтернатив 2, 3, 12 (6 вариантов). Последние шесть позиций
занимают все перестановки альтернатив 1, 4, 5, 6, 7, 13 (720 вариантов). Итого --
4320 варианта оптимальных расположений тринадцати альтернатив для заданных десяти
ранжирований.

Поскольку это не первый наблюдаемый случай получения не всех медиан Кемени
с помощью метода ветвей и границ (Branch&Bound), создаётся впечатление, что
либо реализация алгоритма не слишком точна, либо сам алгоритм Branch&Bound
не в состоянии найти все медианы Кемени...